core/quorum/vote_strategy.go

```diff
@@ -1,99 +1,168 @@
-	deadlineTimer *time.Timer
+	mux                        *event.TypeMux
+	minBlockTime, maxBlockTime int // min and max block creation deadline
+	minVoteTime, maxVoteTime   int // min and max block voting deadline
+	activeMu                   sync.Mutex
+	blockCreateActive          bool
+	votingActive               bool
+	voteTimer                  *time.Timer
+	deadlineTimer              *time.Timer
+	rand                       *rand.Rand
 }
 
 // NewRandomDeadelineStrategy returns a block maker strategy that
 // generated blocks randomly between the given min and max seconds.
-func NewRandomDeadelineStrategy(mux *event.TypeMux, min, max uint) *randomDeadlineStrategy {
-	if min > max {
-		min, max = max, min
+func NewRandomDeadelineStrategy(mux *event.TypeMux, minBlockTime, maxBlockTime, minVoteTime, maxVoteTime uint, activateVoting, activateBlockCreation bool) *randomDeadlineStrategy {
+	if minBlockTime > maxBlockTime {
+		minBlockTime, maxBlockTime = maxBlockTime, minBlockTime
+	}
+	if minBlockTime == 0 {
+		glog.Info("Set minimum block deadeline interval to 1 second")
+		minBlockTime += 1
+	}
+	if minBlockTime == maxBlockTime {
+		maxBlockTime += 1
 	}
-	if min == 0 {
+
+	if minVoteTime > maxVoteTime {
+		minVoteTime, maxVoteTime = maxVoteTime, minVoteTime
+	}
+	if minVoteTime == 0 {
 		glog.Info("Set minimum block deadeline interval to 1 second")
-		min += 1
+		minVoteTime += 1
 	}
-	if min == max {
-		max += 1
+	if minVoteTime == maxVoteTime {
+		maxVoteTime += 1
 	}
+
+	seed, err := crand.Int(crand.Reader, big.NewInt(math.MaxInt64))
+	if err != nil {
+		glog.Fatalf("Unable to seed (psuedo) random deadline strategy")
+	}
+
 	s := &randomDeadlineStrategy{
-		mux:    mux,
-		min:    int(min),
-		max:    int(max),
-		active: true,
+		mux:               mux,
+		minBlockTime:      int(minBlockTime),
+		maxBlockTime:      int(maxBlockTime),
+		minVoteTime:       int(minVoteTime),
+		maxVoteTime:       int(maxVoteTime),
+		blockCreateActive: activateBlockCreation,
+		votingActive:      activateVoting,
+		rand:              rand.New(rand.NewSource(seed.Int64())),
 	}
+
 	return s
 }
 
-func resetBlockMakerTimer(t *time.Timer, min, max int) {
+func resetTimer(t *time.Timer, duration time.Duration) {
 	t.Stop()
 	select {
 	case <-t.C:
 	default:
 	}
-	t.Reset(time.Duration(min+rand.Intn(max-min)) * time.Second)
+	t.Reset(duration)
 }
 
 // Start generating block create request events.
 func (s *randomDeadlineStrategy) Start() error {
 	if glog.V(logger.Debug) {
-		glog.Infof("Random deadline strategy configured with min=%d, max=%d", s.min, s.max)
+		glog.Infof("Random deadline strategy configured with minBlockTime=%d, maxBlockTime=%d, minVoteTime=%d, maxVoteTime=%d",
+			s.minBlockTime, s.maxBlockTime, s.minVoteTime, s.maxVoteTime)
 	}
-	s.deadlineTimer = time.NewTimer(time.Duration(s.min+rand.Intn(s.max-s.min)) * time.Second)
+
+	s.voteTimer = time.NewTimer(time.Duration(s.minBlockTime+rand.Intn(s.maxVoteTime-s.minVoteTime)) * time.Second)
```

Comment:
@bas-vk based on my earlier suggestion, can you please create the timer for voting and block making only if the node is configured with votingActive or blockCreateActive? I don't see a reason why the timer should be active if the node can't vote or create blocks.